Add comprehensive testscript for `gh ruleset`

andyfeller opened the pull request:
<!--
  Thank you for contributing to GitHub CLI!
  To reference an open issue, please write this in your description: `Fixes #NUMBER`
-->

This PR adds a comprehensive testscript around `gh ruleset` commandset.

Due to the complexities of the create repository ruleset endpoint fields, this needed to build a JSON file modeling the input values in order to test the subcommands.

```shell
set -o pipefail && GH_ACCEPTANCE_TOKEN=$(gh auth token --hostname staffship-01.ghe.com) GH_ACCEPTANCE_HOST=staffship-01.ghe.com GH_ACCEPTANCE_ORG=gh-acceptance-testing go test -tags acceptance -json -run ^TestRulesets$ github.com/cli/cli/v2/acceptance  | tparse --all go test
```

resulting in:

```
┌─────────────────────────────────────────────────────────────────────────────────┐
│  STATUS │ ELAPSED │          TEST           │             PACKAGE               │
│─────────┼─────────┼─────────────────────────┼───────────────────────────────────│
│  PASS   │    8.76 │ TestRulesets/commandset │ github.com/cli/cli/v2/acceptance  │
│  PASS   │    0.00 │ TestRulesets            │ github.com/cli/cli/v2/acceptance  │
└─────────────────────────────────────────────────────────────────────────────────┘
┌────────────────────────────────────────────────────────────────────────────────────┐
│  STATUS │ ELAPSED │             PACKAGE              │ COVER │ PASS │ FAIL │ SKIP  │
│─────────┼─────────┼──────────────────────────────────┼───────┼──────┼──────┼───────│
│  PASS   │  9.28s  │ github.com/cli/cli/v2/acceptance │  --   │  2   │  0   │  0    │
└────────────────────────────────────────────────────────────────────────────────────┘
```

andyfeller committed:
Add comprehensive testscript for `gh ruleset`

cliAutomation commented:
Hi! Thanks for the pull request. Please ensure that this change is linked to an issue by mentioning an issue number in the description of the pull request. If this pull request would close the issue, please put the word 'Fixes' before the issue number somewhere in the pull request body. If this is a tiny change like fixing a typo, feel free to ignore this message.

jtmcg commented on acceptance/acceptance_test.go:
```diff
@@ -126,6 +126,15 @@ func TestLabels(t *testing.T) {
 	testscript.Run(t, testScriptParamsFor(tsEnv, "label"))
 }
 
+func TestRulesets(t *testing.T) {
```
Following the convention of the other commands, let's change this to reflect the command name 🙂 

```suggestion
func TestRuleset(t *testing.T) {
```

andyfeller replied:
@jtmcg : are you sure this is a convention?

https://github.com/cli/cli/blob/44ee17760709bedf1cad7e452e61751489b81a33/acceptance/acceptance_test.go#L30

https://github.com/cli/cli/blob/44ee17760709bedf1cad7e452e61751489b81a33/acceptance/acceptance_test.go#L39

https://github.com/cli/cli/blob/44ee17760709bedf1cad7e452e61751489b81a33/acceptance/acceptance_test.go#L48

https://github.com/cli/cli/blob/44ee17760709bedf1cad7e452e61751489b81a33/acceptance/acceptance_test.go#L75

https://github.com/cli/cli/blob/44ee17760709bedf1cad7e452e61751489b81a33/acceptance/acceptance_test.go#L84

https://github.com/cli/cli/blob/44ee17760709bedf1cad7e452e61751489b81a33/acceptance/acceptance_test.go#L102

https://github.com/cli/cli/blob/44ee17760709bedf1cad7e452e61751489b81a33/acceptance/acceptance_test.go#L111

https://github.com/cli/cli/blob/44ee17760709bedf1cad7e452e61751489b81a33/acceptance/acceptance_test.go#L120

https://github.com/cli/cli/blob/44ee17760709bedf1cad7e452e61751489b81a33/acceptance/acceptance_test.go#L129

https://github.com/cli/cli/blob/44ee17760709bedf1cad7e452e61751489b81a33/acceptance/acceptance_test.go#L138

jtmcg replied:
Gah, you're right. I crossed the streams from [this review comment](https://github.com/cli/cli/pull/9783#discussion_r1806724709) on my `repo` PR about the _directory name_, not the test name itself. I guess that means `TestRepo` is the only one not pluralized, now 😅

jtmcg replied:
Nope, it isn't. We're like 50/50 plural vs singular

williammartin replied:
I decided I didn't want to address this on every PR, sorry that I lead you astray for this review. I have a general idea to forgo all the commands so that it's one test with a table so it's just a lowercase string version of the command so we don't go round on this every time.

```go
func TestAcceptance(t *testing.T) {
   tests := []string{ "pr", "issue", ...}
   // loop and subtest
}

jtmcg commented on acceptance/testdata/ruleset/commandset.txtar:
```diff
@@ -0,0 +1,62 @@
+# Setup environment variables used for testscript
```
I think we should follow the same convention for this file's name as we have for the other files: `ruleset-list-check`

andyfeller replied:
I'm open to discuss as all of the other major primitive testscripts were more thoroughly designed into discrete testscripts for specific capabilities.

This testscript is an encompasses testing various behaviors of the whole commandset; there is no single focus.  Ideally, we'd have actual separate testscripts that are specific to subcommands.

But let's go look at the other testscripts:

**Fine-grain scenario testscripts**

- `issue`: https://github.com/cli/cli/tree/trunk/acceptance/testdata/issue
- `pr`: https://github.com/cli/cli/tree/trunk/acceptance/testdata/pr
- `release`: https://github.com/cli/cli/tree/trunk/acceptance/testdata/release
- `repo`: https://github.com/cli/cli/tree/trunk/acceptance/testdata/repo
- `secret`: https://github.com/cli/cli/tree/trunk/acceptance/testdata/secret
- `variable`: https://github.com/cli/cli/tree/trunk/acceptance/testdata/variable
- `workflow`, `run`: https://github.com/cli/cli/tree/trunk/acceptance/testdata/workflow
- 

**Coarse-grain comprehensive testscripts**

- `gpg-key`: https://github.com/cli/cli/blob/trunk/acceptance/testdata/gpg-key/gpg-key.txtar
- `label`: https://github.com/cli/cli/blob/trunk/acceptance/testdata/label/label.txtar
- `ssh-key`: https://github.com/cli/cli/blob/trunk/acceptance/testdata/ssh-key/ssh-key.txtar

Being more coarse grained, I should rename this to `extension.txtar` being it isn't a collection of fine-grained testscripts.

jtmcg replied:
🤔 I'm not sure I agree with these course-grain names...

I would call [gpg-key](https://github.com/cli/cli/blob/trunk/acceptance/testdata/gpg-key/gpg-key.txtar) and [ssh-key](https://github.com/cli/cli/blob/trunk/acceptance/testdata/ssh-key/ssh-key.txtar) "comprehensive" testscripts, as they contains all the subcommands in the gpg-key and ssh-key command trees (`add`, `delete`, `list`)

I wouldn't call `label` or `ruleset` comprehensive, though:

- [label](https://github.com/cli/cli/blob/trunk/acceptance/testdata/label/label.txtar) is missing `clone` and `delete`
- ruleset (this PR) is missing `view`

Either way, I had to click into each of these and scan the code to verify which subcommands were covered. To me, it's tidier to be explicit and consistent, regardless of whether or not we're comprehensively covered.

There is the marginal benefit, too, that if we add another subcommand to any of these command sets we'll probably do this renaming again later

jtmcg replied:
I guess I don't think we should make the distinction between "course-grained" and "fine-grained" test scripts. They're just test scripts.

jtmcg replied:
To close this out, @andyfeller and I discussed on Zoom and agree its fine to leave as-is for now

jtmcg reviewed (commented):
Looks good but for a few name changes 🙂

andyfeller committed:
Rename testscript for consistency

Similar to `gpg-key`, `label`, `ssh-key`, this coarse grained testscript should be named after the commandset given it isn't a collection of targeted scenarios.

jtmcg reviewed (approved):
⭐

williammartin commented on acceptance/testdata/ruleset/ruleset.txtar:
```diff
@@ -0,0 +1,62 @@
+
+# Verify repository ruleset associated with branch
+exec gh ruleset check
+stdout '- pull_request:.+dismiss_stale_reviews_on_push: false.+require_code_owner_review: true.+require_last_push_approval: false.+required_approving_review_count: 1.+required_review_thread_resolution: false'
```
Wow the `ruleset check` output leaves a bit to be desired 😅

andyfeller replied:
Yeah ... wasn't enthused about that ... also the whole GitHub plan piece you mentioned above.

williammartin reviewed (approved):
Nice work on the API request.

Just noting that this doesn't work out of the box for me on dotcom, erroring with:

```
Upgrade to GitHub Pro or make this repository public to enable this feature.
```

But I'm content that this has tackled the primary purpose of the test, thanks!

williammartin commented:
@andyfeller I rebased this on `trunk` and force pushed to fix the merge conflict. I figured it probably wouldn't cause you any issues since the branch will be merged. Hope that's ok!
